pkg/controller/sync/updater.go

```diff
@@ -0,0 +1,222 @@
+	return u.operationCount == 0
+}
+
+func (u *federatedUpdaterImpl) Wait() (map[string]string, bool) {
+	versions := make(map[string]string)
+	ok := true
+
+	timedOut := false
+	start := time.Now()
+	for i := 0; i < u.operationCount; i++ {
+		now := time.Now()
+		if !now.Before(start.Add(u.timeout)) {
+			timedOut = true
+			break
+		}
+		select {
+		case result := <-u.resultChan:
+			if !result.ok {
+				ok = false
+				break
+			}
+			if len(result.version) > 0 {
+				versions[result.clusterName] = result.version
+			}
+		case <-time.After(start.Add(u.timeout).Sub(now)):
+			timedOut = true
+			break
+		}
+	}
+	if timedOut {
+		ok = false
+		u.fedResource.RecordError("OperationTimeoutError", errors.Errorf("Failed to finish all operations in %v", u.timeout))
+	}
+
+	return versions, ok
+}
+
+func (u *federatedUpdaterImpl) Create(clusterName string) {
+	u.operationCount += 1
+	const op = "create"
+	go u.clusterOperation(clusterName, op, func(client util.ResourceClient) (string, error) {
+		u.recordEvent(clusterName, op, "Creating")
+
+		obj, err := u.fedResource.ObjectForCluster(clusterName)
+		if err != nil {
+			return "", err
+		}
+		createdObj, err := client.Resources(obj.GetNamespace()).Create(obj, metav1.CreateOptions{})
+		if err != nil {
+			return "", err
+		}
+		return util.ObjectVersion(createdObj), err
+	})
+}
+
+func (u *federatedUpdaterImpl) Update(clusterName string, clusterObj *unstructured.Unstructured) {
+	u.operationCount += 1
+	const op = "update"
+	go u.clusterOperation(clusterName, op, func(client util.ResourceClient) (string, error) {
+		obj, err := u.fedResource.ObjectForCluster(clusterName)
+		if err != nil {
+			return "", err
+		}
+		err = RetainClusterFields(u.fedResource.TargetKind(), obj, clusterObj, u.fedResource.Object())
+		if err != nil {
+			wrappedErr := errors.Wrapf(err, "Failed to retain fields from %s %q for cluster %q",
+				u.fedResource.FederatedKind(), u.fedResource.FederatedName(), clusterName)
+			return "", wrappedErr
+		}
+
+		version, err := u.fedResource.VersionForCluster(clusterName)
+		if err != nil {
+			return "", err
+		}
+		if !util.ObjectNeedsUpdate(obj, clusterObj, version) {
+			// Resource is current
+			return "", nil
+		}
+
+		// Only record an event if the resource is not current
+		u.recordEvent(clusterName, op, "Updating")
+
+		updatedObj, err := client.Resources(obj.GetNamespace()).Update(obj, metav1.UpdateOptions{})
+		if err != nil {
+			return "", err
+		}
+		return util.ObjectVersion(updatedObj), nil
+	})
+}
+
+func (u *federatedUpdaterImpl) Delete(clusterName string) {
+	u.operationCount += 1
+	const op = "delete"
+	go u.clusterOperation(clusterName, op, func(client util.ResourceClient) (string, error) {
+		u.recordEvent(clusterName, op, "Deleting")
+
+		qualifiedName := u.fedResource.TargetName()
+		orphanDependents := false
+		err := client.Resources(qualifiedName.Namespace).Delete(qualifiedName.Name, &metav1.DeleteOptions{OrphanDependents: &orphanDependents})
```

Comment:
`DeleteOptions` works but it is preferred to use `PropagationPolicy` instead.

Reply:
Good catch.  This really shouldn't be provided at all.  Removed.